pkg/deployment/cert.go

```diff
@@ -0,0 +1,176 @@
+	"github.com/openstack-k8s-operators/lib-common/modules/common/helper"
+	"github.com/openstack-k8s-operators/lib-common/modules/common/secret"
+)
+
+// GenerateTLSCerts generates  a secret containing all the certificates for the relevant service
+// This secret will be mounted by the ansibleEE pod as an extra mount when the service is deployed.
+func GenerateTLSCerts(ctx context.Context, helper *helper.Helper,
+	instance *dataplanev1.OpenStackDataPlaneNodeSet,
+	allIPSets map[string]infranetworkv1.IPSet, serviceName string) error {
+
+	certsData := map[string][]byte{}
+
+	// for each node in the nodeset, issue all the TLS certs needed based on the
+	// ips or DNS Names
+	for nodeName := range instance.Spec.Nodes {
+		var dnsNames []string
+		var secretName string
+		var certName string
+		var certSecret *corev1.Secret = nil
+		var err error
+
+		// TODO(alee) decide if we want to use other labels
+		// For now we just add the hostname so we can select all the certs on one node
+		labels := map[string]string{
+			"hostname": nodeName,
+		}
+
+		ipSet, ok := allIPSets[nodeName]
+		if ok {
+			for _, res := range ipSet.Status.Reservation {
+				fqdnName := strings.Join([]string{nodeName, res.DNSDomain}, ".")
+				dnsNames = append(dnsNames, fqdnName)
+			}
+		}
+
+		switch serviceName {
+		default:
+			// The default case provides a cert with all the dns names for the host.
+			// This will probably be sufficient for most services.  If a service needs
+			// a different kind of cert (for example, containing ips, or using a different
+			// issuer)  then add a case for the service in this switch statement
+
+			secretName = "cert-" + nodeName
+			certSecret, _, err = secret.GetSecret(ctx, helper, secretName, instance.Namespace)
+			if err != nil {
+				if !k8serrors.IsNotFound(err) {
+					err = fmt.Errorf("Error retrieving secret %s - %w", secretName, err)
+					return err
+				}
+
+				certName = secretName
+				certSecret, err = IssueTLSCert(ctx, helper, certName, instance.Namespace,
+					secretName, nodeName, dnsNames, nil, RootCAIssuerInternalLabel,
+					RootCAIssuerInternalNamespaceLabel, labels)
+				if err != nil {
+					return err
+				}
+			}
+		}
+
+		// TODO(alee) Add an owner reference to the secret so it can be monitored
+		// We'll do this once stuggi adds a function to do this in libcommon
+
+		// To use this cert, add it to the relevant service data
+		for key, value := range certSecret.Data {
+			certsData[nodeName+"-"+key] = value
+		}
+	}
+
+	// create a secret to hold the certs for the service
+	serviceCertsSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      GetServiceCertsSecretName(instance, serviceName),
+			Namespace: instance.Namespace,
+		},
+		Data: certsData,
+	}
+	_, _, err := secret.CreateOrPatchSecret(ctx, helper, instance, serviceCertsSecret)
+	if err != nil {
+		err = fmt.Errorf("Error creating certs secret for %s - %w", serviceName, err)
+		return err
+	}
```

Comment:
wondering if we have to create a single with the certs, or just could mount certs from several cert secrets based on the label using the `secretTamplate` as in https://cert-manager.io/docs/usage/certificate/ . would just have to update https://github.com/openstack-k8s-operators/lib-common/blob/main/modules/certmanager/certificate.go#L167 for having the `secretTemplate`

Reply:
In my opinion when we pursue the second option, then it means that we can mount only the certs that are needed. Which is in general safer approach.  The `serviceCertsSecret` aggregates all the individual node certificates into one secret, is this correct? Having one secret with all certificates for a service across all nodes is a point of potential issue.

Reply:
I submitted https://github.com/openstack-k8s-operators/lib-common/pull/358 so that annotations/labels can be passed for the `secretTemplate`

Reply:
As I understand the way things work, there is a single ansibleEE pod that is created for a service, which installs the service on every node in the nodeset.  That means that if there are 1000 nodes in the nodeset, then there are 1000 secrets to be mounted.  Is there a limit to the number of mounts that a pod can have?  Does it make sense to have 1000 mounts?   I created this single secret because I was concerned with not breaching that limit.  

But, I also realize that there is a limit in the size of the secret (4M?)  So, how to avoid that limit?

* We can also create multiple secrets (with as many certs as can fit in a secret) - and then just mount all these secrets as volumes. 

* Another alternative is to create a single volume per service and copy all the certs and keys into that volume.  The main concern there though is that this volume will contain a bunch of secret stuff - ie. keys.  We can specify that this volume is ephemeral and should be deleted - but we don't have any control of when that happens.
 
So, I'm looking for guidance here.  What is the expectation of the size of a nodeset?

Reply:
Based on some back of the envelope calculations, I figure we can end up with about 500 certs in a single secret.
That seems reasonable as an approach.

We do need to do some work where we create multiple secrets if we go over that amount -- well, really keep track of the secret size as we add certs - but I'd like to leave that to a new PR.

I will look into only removing the cacerts from the secret once the PRs with the cacert chains are merged.

Reply:
When we rotate a cert for a node, we should only touch its own secret. To avoid bind mounting housands of certs in large envs, let's use kolla to copy it in instead. This is how we could mitigate mounts count and secret size limits

Reply:
@bogdando We're using a secret containing a bunch of other secrets here rather than a bind-mounted directory because we're concerned that the directory could stick around awhile after the ansible EE pod is completed, potentially exposing the secrets for all the compute nodes.  Secrets have some built-in protections.  They are stored encrypted in etcd, so they are safe at rest.  They are only sent to the nodes that need them and they are mounted using ramfs - so there is no place where they are written to disk.  The secret size limitation is because the data is stored in etcd.

We could potentially use a mounted  volume if that volume were encrypted in some way, but as far as I can tell, there is no openstack-style encrypted cinder volume available in k8s, where a volume is mounted with an encryptor/decryptor mechanism.  If you know of such a mechanism, I'd be happy to use it.  I did find mention of something like that here -- https://docs.portworx.com/portworx-enterprise/operations/key-management/kubernetes-secrets/pvc-encryption-using-annotations -- but not in openshift.

For openshift, there appear to be a couple of options. 
https://access.redhat.com/documentation/en-us/red_hat_openshift_container_storage/4.6/html/planning_your_deployment/security-considerations_rhocs (configuring encryption at the container storage level)

https://red-hat-storage.github.io/ocs-training/training/ocs4/ocs4-encryption.html#_overview (integration with vault as a kms)

All of this is way beyond the scope of this patch :)

For the reconciliation loop issue, recall that there will be maybe 500 certs in a secret, so the number of extra reconciliations will be somewhat small.  Also, we can add a label to the secret of secrets and some code to tell the  reconciliation loop to ignore it.  We can handle that in a follow on patch when we test reconciations for cert changes.

 I propose we go with this mechanism for now - we know it works - and it provides the needed security without too much extra complexity.